### PR TITLE
Account derivation feedback

### PR DESCRIPTION
**Pull request by hbulgarini**
PR related to issue : #22

### CODE REVIEW

File: src/App.tsx
Comment (wirednkod) [inline src/App.tsx:29]: I would not move the Main page inside the components. I normally prefer a distinguish between reusable components and page-components.

File: src/actions/accountActions.ts
Comment (wirednkod) [inline src/actions/accountActions.ts:21]: are the spaces at the end of each string (SET_DERIVED_ACCOUNT and SET_GENERIC_ACCOUNT intentional? looks like a typo

File: src/actions/transactionActions.ts
Comment (wirednkod) [inline src/actions/transactionActions.ts:26]: Again some spaces at the end of each string. Typo?

File: src/components/AccountDisplay.tsx
Comment (wirednkod) [inline src/components/AccountDisplay.tsx:83]: I'd go with this:
```
<p>{balance || '-'}</p>
```

File: src/components/AccountDisplay.tsx
Comment (wirednkod) [inline src/components/AccountDisplay.tsx:97]: I'd go with this:
```
<p>{balance || '-'}</p>
```

File: src/hooks/useBalance.ts
Comment (wirednkod) [inline src/hooks/useBalance.ts:42]: I do not understand why to remove the double functionality of useBalance that can be very helpful in a project like bridges. Its more convenient to receive both amounts and chose which one to use than calling it twice. Or am I missing something?

File: src/types/accountTypes.ts
Comment (wirednkod) [inline src/types/accountTypes.ts:28]: I would avoid using any if possible

File: src/util/getReceiverAddress.ts
Comment (wirednkod) [inline src/util/getReceiverAddress.ts:59]: You should have also a GENERIC error (in case if  (!receiverAddress)

**Review by wirednkod — CHANGES_REQUESTED** [review]
I would prefer to merge this PR **after** PR #96 if possible

File: src/hooks/useBalance.ts
Comment (hbulgarini) [reply src/hooks/useBalance.ts:42]: the reason why i dissected the logic to have specific calls per chain is because this functionality was quite tight to the select account component use case when you have the input  [addrress, chain], and you get the balances for the account and derived account.  In the `Transfer` component i realized that  i needed this logic extracted so i only have one account with one balance. 
In the previous version `Account` was retrieving always two rows: Account and derived  and in transfers sometimes i just need only one.

**Review by wirednkod — APPROVED** [review]

File: src/hooks/usePrevious.ts
Comment (tomusdrw) [inline src/hooks/usePrevious.ts:25]: This thing looks super hacky, why do we `useRef`? Can't we have an explicit state with previous value (cause I believe that's what this thing is supposed to do).

File: src/util/getReceiverAddress.ts
Comment (tomusdrw) [inline src/util/getReceiverAddress.ts:31]:
```suggestion
 const [validatedDerivedAcccount, rest] = checkAddress(receiverAddress, SS58Format);
    const address = receiverAddress;
    if (validatedDerivedAcccount) {
       return { address, formatFound: chain };
    }
    // should be extracted as a separate component/function
    const getFormat = (prefix: string) => {
       const intPrefix: number = parseInt(prefix, 10);
       if (intPrefix === GENERIC_SUBSTRATE_PREFIX /* 42 */) {
          return GENERIC;
       }
       const chainsConfigs = getChainConfigs();
       return Object.keys(chainsConfigs).find((key) => chainsConfigs[key].SS58Format === intPrefix);
    }
    
    const parts = rest?.split(',');
    const prefix = parts![2].split(' ');
    const formatFound = getFormat(prefix);
    
    const address = getDeriveAccount({
        SS58Format,
        address,
        bridgeId
      });

      return { address, formatFound };
    }
```

The whole function could be made more readable.

File: src/components/Receiver.tsx
Comment (tomusdrw) [inline src/components/Receiver.tsx:63]: IMHO to unify the way accounts are displayed we should always display the bottom part. In case it's just target chain format we can repeat the address.

Comment (tomusdrw) [reply src/components/Receiver.tsx:63]: Also right now we don't handle correctly accounts with prefixes being neither source, target or generic.

File: src/components/ReceiverInput.tsx
Comment (tomusdrw) [inline src/components/ReceiverInput.tsx:115]: We shouldn't be assuming that it's sourceChain account if `formatFound` is neither `GENERIC` nor `targetChain`.

**Review by tomusdrw — CHANGES_REQUESTED** [review]
It's much better now, but there is a lot of details that need improvement. Some of them are minor things, but some are pretty critical.
The components logic seems to be overcomplicated as well and when I tested the UI I can get it into weird states, like:
derived accounts being clearly wrong (either in sender selector or receiver selector):
image

Clicking "Native" ends up with "derived" (so does clicking "Companion"):
image

Address being neither source nor target nor generic - no idea what address will receive funds:
image

Derived DAVE showing 0 balance after selection ,but has balance durign the selection:
image
image

Able to send transaction with address being detected as invalid (balance can be left empty too).
image

File: src/hooks/usePrevious.ts
Comment (wirednkod) [reply src/hooks/usePrevious.ts:25]: Actually this hook is the best way to retrieve the previous value without having a second state to keep the value (plus its reusable for multiple components - so I would leave it as is.

**Review by wirednkod — APPROVED** [review]

File: src/components/Receiver.tsx
Comment (hbulgarini) [reply src/components/Receiver.tsx:63]: actually that brought some hard restrictions when i needed to implement the receiver component. Having a component that always retrieve two lines makes difficult to implement different use cases as the generic case for example.

File: src/components/ReceiverInput.tsx
Comment (hbulgarini) [reply src/components/ReceiverInput.tsx:115]: why not? what are the other options then?

File: src/components/Sender.tsx
Comment (hbulgarini) [reply src/components/Sender.tsx:83]: i think this can be addressed after ... do you see it as a hard requirement?

File: src/hooks/usePrevious.ts
Comment (hbulgarini) [reply src/hooks/usePrevious.ts:25]: yes, actually this is quite used and there is a lot of  examples out there.

File: src/components/ReceiverDerivedAccount.tsx
Comment (tomusdrw) [inline src/components/ReceiverDerivedAccount.tsx:23]: Let's remove if unused.

File: src/components/GenericAccount.tsx
Comment (tomusdrw) [inline src/components/GenericAccount.tsx:53]: Please remove if this is unused.

File: src/components/AccountDisplay.tsx
Comment (tomusdrw) [inline src/components/AccountDisplay.tsx:80]: Seems like two completelty different components to me (balance display & account display), but let's leave it like that for now.

File: src/components/AccountDisplay.tsx
Comment (tomusdrw) [inline src/components/AccountDisplay.tsx:83]: I'd add a `title/tooltip` with full account address, can be left for separate PR though.

File: src/components/ReceiverInput.tsx
Comment (tomusdrw) [reply src/components/ReceiverInput.tsx:115]: Yes, there is 16k possible SS58 prefix values, and we should only handle 3 of them: `42 (generic)`, source, target.

Comment (tomusdrw) [reply src/components/ReceiverInput.tsx:115]: Still something to fix.

File: src/components/Sender.tsx
Comment (tomusdrw) [inline src/components/Sender.tsx:71]: `fromSender` just displays balance, why bother with name and `isDerived`?

File: src/components/Sender.tsx
Comment (tomusdrw) [reply src/components/Sender.tsx:83]: Sure, please create an issue for this.

File: src/components/Transfer.tsx
Comment (tomusdrw) [inline src/components/Transfer.tsx:73]:
```suggestion
       Send Bridge Message 
```
Current label is super confusing, the transfer is NOT between networks.

File: src/hooks/useApiBalance.ts
Comment (tomusdrw) [inline src/hooks/useApiBalance.ts:53]:
```suggestion
  const addressResult = !isDerived ? address : getDeriveAccount({
      SS58Format,
      address,
      bridgeId
    });
```

File: src/util/getReceiverAddress.ts
Comment (tomusdrw) [inline src/util/getReceiverAddress.ts:38]:
```suggestion
      if (intPrefix === GENERIC_SUBSTRATE_PREFIX) {
```
I was just suggesting that the constant should contain that value.

File: src/components/AccountIdenticon.tsx
Comment (tomusdrw) [inline src/components/AccountIdenticon.tsx:29]: I still think the placeholder should be grayscale to make it look even less like actual identicon.
```suggestion
    opacity: '30%', // as a fallback if filter is not supported
    filter: 'grayscale(1)'
```

File: src/components/ReceiverDerivedAccount.tsx
Comment (tomusdrw) [inline src/components/ReceiverDerivedAccount.tsx:60]: IMHO to unify the display of the component we should always render the bottom part and just repeat the same stuff that is in the input field, but that's more up to @goldsteinsveta. AFAIR on designs we do display `raw/derived` always and while I already mentioned I think we can still improve on naming, but at least presentation wise I think it was good.

**Review by tomusdrw — APPROVED** [review]
looks much better now! A bunch of smaller things to address, but can be merged after that.

File: src/components/Transfer.tsx
Comment (hbulgarini) [reply src/components/Transfer.tsx:73]: it was actually on the design ... i can change it

File: src/components/GenericAccount.tsx
Comment (hbulgarini) [reply src/components/GenericAccount.tsx:53]: i prefer this component to stay even if it is not used. Removing it will lead us to work on something that needs adjustments not recreating the component from scratch.

File: src/components/ReceiverDerivedAccount.tsx
Comment (hbulgarini) [reply src/components/ReceiverDerivedAccount.tsx:23]: i prefer this component to stay even if it is not used. Removing it will lead us to work on something that needs adjustments not recreating the component from scratch.

File: src/components/GenericAccount.tsx
Comment (tomusdrw) [reply src/components/GenericAccount.tsx:53]: No, please remove. It will make it harder to review in the future and dead code should not be in the repository.

File: src/components/Sender.tsx
Comment (hbulgarini) [reply src/components/Sender.tsx:83]: https://github.com/paritytech/parity-bridges-ui/issues/103

File: src/components/AccountDisplay.tsx
Comment (hbulgarini) [reply src/components/AccountDisplay.tsx:80]: Right , it was not hard to create it now.

File: src/components/AccountDisplay.tsx
Comment (hbulgarini) [reply src/components/AccountDisplay.tsx:83]: https://github.com/paritytech/parity-bridges-ui/issues/104

File: src/components/ReceiverInput.tsx
Comment (hbulgarini) [reply src/components/ReceiverInput.tsx:115]: Done.

File: src/components/Sender.tsx
Comment (hbulgarini) [reply src/components/Sender.tsx:71]: Completely removed.

**Review by hbulgarini — COMMENTED** [review]
In the last commits i'm addressing all your comments and also i'm solving the issue with the derived account logic.

File: src/components/Sender.tsx
Comment (hbulgarini) [reply src/components/Sender.tsx:106]: Your code has an issue and this is why i reverted it to the original one: 

When you flip the chains because a target account was selected, the order in the dropdown is flipped as well.

**Comment by hbulgarini** [comment]
know issue: https://github.com/paritytech/parity-bridges-ui/issues/105

**Comment by hbulgarini** [comment]
> It's much better now, but there is a lot of details that need improvement. Some of them are minor things, but some are pretty critical.
> The components logic seems to be overcomplicated as well and when I tested the UI I can get it into weird states, like:
> derived accounts being clearly wrong (either in sender selector or receiver selector):
> image
> 
> Clicking "Native" ends up with "derived" (so does clicking "Companion"):
> image
> 
> Address being neither source nor target nor generic - no idea what address will receive funds:
> image
> 
> Derived DAVE showing 0 balance after selection ,but has balance durign the selection:
> image
> image
> 
> Able to send transaction with address being detected as invalid (balance can be left empty too).
> image

In the PR i have solved already in this PR the following issues:

- Incorrect derived accounts (now all derived accounts shows the same as the CLI)
- Balances reflected as part of the fix described above.
- GenericAccount component removed and to be implemented in issue #101 
- Major improvements / simplification to components.

**Pull request merged**
